File: Data Structures & Algorithms/is-palindrome/submission-0.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        left, right = 0, len(s) - 1

        while left < right:
            if not self.is_alphanumeric(s[left]):
                left += 1
                continue
            if not self.is_alphanumeric(s[right]):
                right -= 1
                continue
            
            if s[left].lower() != s[right].lower():
                return False
            
            left += 1
            right -= 1
        
        return True
        
    def is_alphanumeric(self, s: str) -> bool:
        return (
            ord('A') <= ord(s) <= ord('Z')
            or ord('a') <= ord(s) <= ord('z')
            or ord('0') <= ord(s) <= ord('9')
        )

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-0.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//     pub val: i32,
//     pub left: Option<Rc<RefCell<TreeNode>>>,
//     pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//     #[inline]
//     pub fn new(val: i32) -> Self {
//         TreeNode {
//             val,
//             left: None,
//             right: None,
//         }
//     }
// }

use std::cell::RefCell;
use std::rc::Rc;

impl Solution {
    pub fn kth_smallest(root: Option<Rc<RefCell<TreeNode>>>, k: i32) -> i32 {
        fn dfs(root: &Option<Rc<RefCell<TreeNode>>>, nums: &mut Vec<i32>) {
            match root {
                None => {}
                Some(node) => {
                    dfs(&node.borrow().left, nums);
                    nums.push(node.borrow().val);
                    dfs(&node.borrow().right, nums);
                }
            }
        }
        let mut nums = vec![];
        dfs(&root, &mut nums);
        nums[k as usize - 1]
    }
}
